master/internal/restore.go

```diff
@@ -134,79 +133,68 @@ func (m *Master) restoreExperiment(expModel *model.Experiment) error {
 	return nil
 }
 
-// restoreTrial takes the a searcher.Create and attempts to restore the trial that would be
-// associated with it. On failure, the trial is just reset to the start and errors are logged.
-func (e *internalExperiment) restoreTrial(
-	ckpt *model.Checkpoint, searcher experiment.TrialSearcherState,
+func (e *internalExperiment) restoreRun(
+	ckpt *model.Checkpoint, searcher experiment.RunSearcherState,
 ) {
-	l := e.syslog.WithField("request-id", searcher.Create.RequestID)
-	l.Debug("restoring trial")
+	syslog := e.syslog
+	syslog.Debug("restoring run")
 
+	var trial *model.Trial
 	var trialID *int
-	var terminal bool
-	switch trial, err := db.TrialByExperimentAndRequestID(context.TODO(),
-		e.ID, searcher.Create.RequestID); {
-	case errors.Cause(err) == db.ErrNotFound:
-		l.Debug("trial was not previously persisted")
-	case err != nil:
-		// This is the only place we _have_ to error, because if the trial did previously exist
-		// and we failed to retrieve it, continuing will result in an invalid state (we'll get a
-		// new trial in the trials table with the same (experiment_id, request_id).
-		l.WithError(err).Error("failed to retrieve trial, aborting restore")
-		terminal = true
-	default:
-		trialID = &trial.ID
-		l = l.WithField("trial-id", trial.ID)
-		if model.TerminalStates[trial.State] {
-			l.Debugf("trial was in terminal state in restore: %s", trial.State)
-			terminal = true
+	var err error
+
+	if searcher.RunID != nil {
```

Comment:
Something doesn't smell right here.

- Why is the runid ever nil on the snapshot?
- Assuming the previous answer is "b/c ordering is hard", then why do we save said snapshot to the db?
- Assuming there is a good answer to that, then why do we create a new trial as part of restoring?  If there was a good reason to have a nil RunID before the shutdown, then why doesn't that reason still apply during restore?

Reply:
yeah, this was confusing to me as well. i was trying my best to refactor existing code while keeping the same behavior. AFAICT, the old code _was_ creating a new trial if the snapshot's trial wasn't created. 

in previous `restore.go`, we try to get trial by request ID. if it doesn't exist, we create a new trial. in `experiment.go`, searcher wants to create trial -> we set the trial state -> create trial -> snapshot. but i agree, i can't imagine why you'd want to recreate the trial if something fails before snapshot. 

maybe this behavior is just a leftover artifact of some refactored behavior, like if searcher create -> snapshot used to be a more async process. happy to get rid of it, but maybe @stoksc can confirm here?